MMCore/MMCore.cpp

```diff
@@ -873,6 +899,107 @@ void CMMCore::initializeAllDevices() throw (CMMError)
    updateCoreProperties();
 }
 
+
+/**
+ * Calls Initialize() method for each loaded device.
+ * This implementation initializes devices on separate threads, one per device module (adapter).
+ * This method also initializes allowed values for core properties, based
+ * on the collection of loaded devices.
+ */
+void CMMCore::initializeAllDevicesParallel() throw (CMMError)
+{
+   std::vector<std::string> devices = deviceManager_->GetDeviceList();
+   LOG_INFO(coreLogger_) << "Will initialize " << devices.size() << " devices";
+   
+   std::map<std::shared_ptr<LoadedDeviceAdapter>, std::deque<std::pair<std::shared_ptr<DeviceInstance>, std::string>>> moduleMap;
+   std::vector<std::shared_ptr<DeviceInstance>> ports;
+
+   // first round, collect all DeviceAdapters
+   for (size_t i = 0; i < devices.size(); i++)
+   {
+      std::shared_ptr<DeviceInstance> pDevice;
+      try {
+         pDevice = deviceManager_->GetDevice(devices[i]);
+      }
+      catch (CMMError& err) {
+         logError(devices[i].c_str(), err.getMsg().c_str());
+         throw;
+      }
+      if (pDevice->GetType() == MM::SerialDevice)
+      {
+         ports.push_back(pDevice);
+      }
+      else {
+         std::shared_ptr<LoadedDeviceAdapter> pAdapter;
+         pAdapter = pDevice->GetAdapterModule();
+
+         if (moduleMap.find(pAdapter) == moduleMap.end())
+         {
+            std::deque<std::pair<std::shared_ptr<DeviceInstance>, std::string>> pDevices;
+            pDevices.push_back(make_pair(pDevice, devices[i]));
+            moduleMap.insert({ pAdapter, pDevices });
+         }
+         else
+         {
+            moduleMap.find(pAdapter)->second.push_back(make_pair(pDevice, devices[i]));
+         }
+      }
+   }
+
+   // Initialize ports first.  This should be gast, so no need to go parallel (also could not hurt really)
+   for (std::shared_ptr<DeviceInstance> pPort : ports)
+   {
+      mm::DeviceModuleLockGuard guard(pPort);
+      LOG_INFO(coreLogger_) << "Will initialize device " << pPort->GetLabel();
+      pPort->Initialize();
+      LOG_INFO(coreLogger_) << "Did initialize device " << pPort->GetLabel();
+   }
+
+   // second round, spin up threads to initialize non-port devices, one thread per module
+   std::vector<std::future<int>> futures;
+   std::map<std::shared_ptr<LoadedDeviceAdapter>, std::deque<std::pair<std::shared_ptr<DeviceInstance>, std::string>>>::iterator it;
+   for (it = moduleMap.begin(); it != moduleMap.end(); it++)
+   {
+      auto f = std::async(std::launch::async, &CMMCore::initializeDequeOfDevices, this, it->second);
+      futures.push_back(std::move(f));
+   }
+   for (auto& f : futures) {
+      // Note: we can do a 'f.wait_for(std::chrono::seconds(20)' to wait up to 20 seconds before giving up
+      // which would avoid hanging with devices that hang in their initialize function
+      f.get();
+   }
```

Comment:
The `std::future` returned by `std::async` is special and its destructor blocks until the future completes. This is okay if there are 0 or 1 errors total (the successful initializations run to completion and the exception is propagated).

When there are 2 or more errors, however, the second exception would be thrown in the destructor of the future, and throwing anything in a destructor is very bad (might terminate by default).

I think it's best to catch the first exception and then explicitly wait for the remaining futures (catching and ignoring any subsequent exceptions). Then the first exception can be rethrown. (`std::current_exception` and `std::rethrow_exception` can be used.)

Reply:
Implemented as you described in a52a350.  Tested with limited hardware (guess we need multiple demo devices that can throw exceptions in the initialize function to fully test this).  

Also tested earlier version on complicated hardware, where it reduced startup times from 32 to 20 seconds, which is an appreciable improvement.